File: backend-spring-boot/src/test/java/no/vicx/backend/calculator/CalculatorSecurityServiceTest.java
package no.vicx.backend.calculator;

import no.vicx.database.calculator.CalculatorRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Set;

import static no.vicx.backend.testconfiguration.SecurityTestUtils.createJwtInTest;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class CalculatorSecurityServiceTest {

    @Mock
    CalculatorRepository calculatorRepository;

    @InjectMocks
    CalculatorSecurityService sut;

    @Test
    void isAllowedToDelete_givenEmptyList_expectTrue() {
        var result = sut.isAllowedToDelete(Collections.emptyList(), TOKEN_IN_TEST);

        assertTrue(result);

        verify(calculatorRepository, never()).findAllIdsByUsername(anyString());
    }

    @Test
    void isAllowedToDelete_givenListWithNull_expectTrue() {
        var result = sut.isAllowedToDelete(Collections.singletonList(null), TOKEN_IN_TEST);

        assertTrue(result);

        verify(calculatorRepository, never()).findAllIdsByUsername(anyString());
    }

    @Test
    void isAllowedToDelete_givenIdsButNoIdsInDatabase_expectFalse() {

        var result = sut.isAllowedToDelete(List.of(1L), TOKEN_IN_TEST);

        assertFalse(result);
    }

    @Test
    void isAllowedToDelete_givenListOfIdsThatDoesNotBelongToUser_expectFalse() {
        var result = sut.isAllowedToDelete(List.of(1L, 2L), TOKEN_IN_TEST);

        assertFalse(result);

        verify(calculatorRepository).findAllIdsByUsername(anyString());
    }

    @Test
    void isAllowedToDelete_givenListOfIdsThatBelongsToUser_expectTrue() {
        when(calculatorRepository.findAllIdsByUsername(anyString()))
                .thenReturn(Set.of(1L));

        var result = sut.isAllowedToDelete(Collections.singletonList(1L), TOKEN_IN_TEST);

        assertTrue(result);

        verify(calculatorRepository).findAllIdsByUsername(anyString());
    }

    @Test
    void isAllowedToDelete_givenListWithValueAndNull_expectTrue() {
        when(calculatorRepository.findAllIdsByUsername(anyString()))
                .thenReturn(Set.of(1L));

        var result = sut.isAllowedToDelete(Arrays.asList(1L, null), TOKEN_IN_TEST);

        assertTrue(result);

        verify(calculatorRepository).findAllIdsByUsername(anyString());
    }

    static final Authentication TOKEN_IN_TEST =
            new JwtAuthenticationToken(createJwtInTest(Collections.singletonList("USER")));
}

File: backend-spring-boot/src/test/kotlin/no/vicx/backend/calculator/CalculatorSecurityServiceTest.kt
package no.vicx.backend.calculator

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.mockk.called
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import no.vicx.backend.SecurityTestUtils
import no.vicx.database.calculator.CalculatorRepository
import org.springframework.security.core.Authentication
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken

class CalculatorSecurityServiceTest : StringSpec({
    val calculatorRepository: CalculatorRepository = mockk()
    val sut = CalculatorSecurityService(calculatorRepository)

    "when calling isAllowedToDelete with empty list of ids then expect true" {
        val result = sut.isAllowedToDelete(
            emptyList(),
            TOKEN_IN_TEST
        )

        result shouldBe true

        verify { calculatorRepository wasNot called }
    }

    "when calling isAllowedToDelete with ids that belongs to another user then expect false" {
        every { calculatorRepository.findAllIdsByUsername(any()) } returns emptySet()

        val result = sut.isAllowedToDelete(listOf(1L, 2L), TOKEN_IN_TEST)

        result shouldBe false
    }

    "when calling isAllowedToDelete with ids that belongs to user then expect true" {
        every { calculatorRepository.findAllIdsByUsername(any()) } returns setOf(1L)

        val result = sut.isAllowedToDelete(listOf(1L), TOKEN_IN_TEST)

        result shouldBe true
    }
}) {
    companion object {
        val TOKEN_IN_TEST: Authentication = JwtAuthenticationToken(
            SecurityTestUtils.createJwtInTest(listOf("USER"))
        )
    }
}